File: api/app/models/category.rb
# frozen_string_literal: true

class Category < ApplicationRecord
  validates :code,        presence: true, uniqueness: true
  validates :title,       allow_empty: true
  validates :description, allow_empty: true
  validates :order,       presence: true

  has_many :problems, dependent: :nullify

  class << self
    # デバッグ用ショートハンド
    def code(code)
      self.find_by(code: code)
    end
  end
end

File: ui/orm/Category.js
import orm from '~/orm'
import BaseModel from '~/orm/BaseModel'

export default class Category extends BaseModel {
  static entity = 'categories'

  static fields() {
    return {
      id: this.string(),
      code: this.string().nullable(),
      title: this.string(),
      description: this.string(),
      order: this.number(),
      problems: this.hasMany(orm.Problem, 'categoryId'),
      updatedAt: this.string(),
    }
  }

  static mutationFields() {
    return {
      code: '',
      title: '',
      description: '',
      order: 0,
    }
  }

  get displayTitle() {
    return $nuxt.isStaff ? `${this.code}. ${this.title}` : this.title
  }
}
